<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.1" />
<title>sake.errors API documentation</title>
<meta name="description" content="The standard error bases which Sake implementations will be raising â€¦" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>sake.errors</code></h1>
</header>
<section id="section-intro">
<p>The standard error bases which Sake implementations will be raising.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These supplement python's builtin exceptions but do not replace them.</p>
</div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-
# cython: language_level=3
# BSD 3-Clause License
#
# Copyright (c) 2020, Faster Speeding
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
#
# * Neither the name of the copyright holder nor the names of its
#   contributors may be used to endorse or promote products derived from
#   this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS &#34;AS IS&#34;
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
&#34;&#34;&#34;The standard error bases which Sake implementations will be raising.

!!! note
    These supplement python&#39;s builtin exceptions but do not replace them.
&#34;&#34;&#34;

from __future__ import annotations

__all__: typing.Final[typing.Sequence[str]] = [
    &#34;BackendError&#34;,
    &#34;CannotDelete&#34;,
    &#34;EntryNotFound&#34;,
    &#34;InvalidDataFound&#34;,
    &#34;SakeException&#34;,
]

import typing


class SakeException(Exception):
    &#34;&#34;&#34;A base exception for the expected exceptions raised by Sake implementations.

    Parameters
    ----------
    message : str
        The exception&#39;s message.
    base : typing.Optional[BaseException]
        The exception which caused this exception if applicable else `builtins.None`.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = (&#34;base_exception&#34;, &#34;message&#34;)

    message: str
    &#34;&#34;&#34;The exception&#39;s message, this may be an empty string if there is no message.&#34;&#34;&#34;

    base: typing.Optional[BaseException]
    &#34;&#34;&#34;The exception which caused this exception if applicable else `builtins.None`.&#34;&#34;&#34;

    def __init__(self, message: str, *, exception: typing.Optional[BaseException] = None) -&gt; None:
        self.message = message
        self.base_exception = exception

    def __repr__(self) -&gt; str:
        return f&#34;{type.__name__}({self.message!r})&#34;


class BackendError(SakeException, ValueError):
    &#34;&#34;&#34;A error raised when communicating with the backend fails

    This may be a sign of underlying network or database issues.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()


class CannotDelete(SakeException, ValueError):  # TODO: implement and document cascading handling?
    &#34;&#34;&#34;An error raised in response to an attempt to delete an entry which can&#39;t be deleted.

    This most likely reason for this to be raised would be due to an attempt to
    deleted a entry that&#39;s being kept alive by references without specifying to
    cascade references in a referential database.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()


class InvalidDataFound(SakeException, LookupError):
    &#34;&#34;&#34;An error raised when the retrieved data is in an unexpected format.

    This may indicate that you are running different versions of a Sake
    implementation with the same database.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()


class EntryNotFound(SakeException, LookupError):
    &#34;&#34;&#34;An error raised in response to an attempt to get an entry which doesn&#39;t exist.

    !!! note
        This shouldn&#39;t ever be raised by a delete method or iter method.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="sake.errors.BackendError"><code class="flex name class">
<span>class <span class="ident">BackendError</span></span>
<span>(</span><span>message:Â str, *, exception:Â typing.Optional[BaseException]Â =Â None)</span>
</code></dt>
<dd>
<div class="desc"><p>A error raised when communicating with the backend fails</p>
<p>This may be a sign of underlying network or database issues.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BackendError(SakeException, ValueError):
    &#34;&#34;&#34;A error raised when communicating with the backend fails

    This may be a sign of underlying network or database issues.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></li>
<li>builtins.ValueError</li>
<li>builtins.Exception</li>
<li>builtins.BaseException</li>
</ul>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></b></code>:
<ul class="hlist">
<li><code><a title="sake.errors.SakeException.base" href="#sake.errors.SakeException.base">base</a></code></li>
<li><code><a title="sake.errors.SakeException.base_exception" href="#sake.errors.SakeException.base_exception">base_exception</a></code></li>
<li><code><a title="sake.errors.SakeException.message" href="#sake.errors.SakeException.message">message</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="sake.errors.CannotDelete"><code class="flex name class">
<span>class <span class="ident">CannotDelete</span></span>
<span>(</span><span>message:Â str, *, exception:Â typing.Optional[BaseException]Â =Â None)</span>
</code></dt>
<dd>
<div class="desc"><p>An error raised in response to an attempt to delete an entry which can't be deleted.</p>
<p>This most likely reason for this to be raised would be due to an attempt to
deleted a entry that's being kept alive by references without specifying to
cascade references in a referential database.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class CannotDelete(SakeException, ValueError):  # TODO: implement and document cascading handling?
    &#34;&#34;&#34;An error raised in response to an attempt to delete an entry which can&#39;t be deleted.

    This most likely reason for this to be raised would be due to an attempt to
    deleted a entry that&#39;s being kept alive by references without specifying to
    cascade references in a referential database.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></li>
<li>builtins.ValueError</li>
<li>builtins.Exception</li>
<li>builtins.BaseException</li>
</ul>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></b></code>:
<ul class="hlist">
<li><code><a title="sake.errors.SakeException.base" href="#sake.errors.SakeException.base">base</a></code></li>
<li><code><a title="sake.errors.SakeException.base_exception" href="#sake.errors.SakeException.base_exception">base_exception</a></code></li>
<li><code><a title="sake.errors.SakeException.message" href="#sake.errors.SakeException.message">message</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="sake.errors.EntryNotFound"><code class="flex name class">
<span>class <span class="ident">EntryNotFound</span></span>
<span>(</span><span>message:Â str, *, exception:Â typing.Optional[BaseException]Â =Â None)</span>
</code></dt>
<dd>
<div class="desc"><p>An error raised in response to an attempt to get an entry which doesn't exist.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This shouldn't ever be raised by a delete method or iter method.</p>
</div></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class EntryNotFound(SakeException, LookupError):
    &#34;&#34;&#34;An error raised in response to an attempt to get an entry which doesn&#39;t exist.

    !!! note
        This shouldn&#39;t ever be raised by a delete method or iter method.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></li>
<li>builtins.LookupError</li>
<li>builtins.Exception</li>
<li>builtins.BaseException</li>
</ul>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></b></code>:
<ul class="hlist">
<li><code><a title="sake.errors.SakeException.base" href="#sake.errors.SakeException.base">base</a></code></li>
<li><code><a title="sake.errors.SakeException.base_exception" href="#sake.errors.SakeException.base_exception">base_exception</a></code></li>
<li><code><a title="sake.errors.SakeException.message" href="#sake.errors.SakeException.message">message</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="sake.errors.InvalidDataFound"><code class="flex name class">
<span>class <span class="ident">InvalidDataFound</span></span>
<span>(</span><span>message:Â str, *, exception:Â typing.Optional[BaseException]Â =Â None)</span>
</code></dt>
<dd>
<div class="desc"><p>An error raised when the retrieved data is in an unexpected format.</p>
<p>This may indicate that you are running different versions of a Sake
implementation with the same database.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class InvalidDataFound(SakeException, LookupError):
    &#34;&#34;&#34;An error raised when the retrieved data is in an unexpected format.

    This may indicate that you are running different versions of a Sake
    implementation with the same database.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = ()</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></li>
<li>builtins.LookupError</li>
<li>builtins.Exception</li>
<li>builtins.BaseException</li>
</ul>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></b></code>:
<ul class="hlist">
<li><code><a title="sake.errors.SakeException.base" href="#sake.errors.SakeException.base">base</a></code></li>
<li><code><a title="sake.errors.SakeException.base_exception" href="#sake.errors.SakeException.base_exception">base_exception</a></code></li>
<li><code><a title="sake.errors.SakeException.message" href="#sake.errors.SakeException.message">message</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="sake.errors.SakeException"><code class="flex name class">
<span>class <span class="ident">SakeException</span></span>
<span>(</span><span>message:Â str, *, exception:Â typing.Optional[BaseException]Â =Â None)</span>
</code></dt>
<dd>
<div class="desc"><p>A base exception for the expected exceptions raised by Sake implementations.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>message</code></strong> :&ensp;<code>str</code></dt>
<dd>The exception's message.</dd>
<dt><strong><code>base</code></strong> :&ensp;<code>typing.Optional[BaseException]</code></dt>
<dd>The exception which caused this exception if applicable else <code>builtins.None</code>.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class SakeException(Exception):
    &#34;&#34;&#34;A base exception for the expected exceptions raised by Sake implementations.

    Parameters
    ----------
    message : str
        The exception&#39;s message.
    base : typing.Optional[BaseException]
        The exception which caused this exception if applicable else `builtins.None`.
    &#34;&#34;&#34;

    __slots__: typing.Sequence[str] = (&#34;base_exception&#34;, &#34;message&#34;)

    message: str
    &#34;&#34;&#34;The exception&#39;s message, this may be an empty string if there is no message.&#34;&#34;&#34;

    base: typing.Optional[BaseException]
    &#34;&#34;&#34;The exception which caused this exception if applicable else `builtins.None`.&#34;&#34;&#34;

    def __init__(self, message: str, *, exception: typing.Optional[BaseException] = None) -&gt; None:
        self.message = message
        self.base_exception = exception

    def __repr__(self) -&gt; str:
        return f&#34;{type.__name__}({self.message!r})&#34;</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li>builtins.Exception</li>
<li>builtins.BaseException</li>
</ul>
<h3>Subclasses</h3>
<ul class="hlist">
<li><a title="sake.errors.BackendError" href="#sake.errors.BackendError">BackendError</a></li>
<li><a title="sake.errors.CannotDelete" href="#sake.errors.CannotDelete">CannotDelete</a></li>
<li><a title="sake.errors.EntryNotFound" href="#sake.errors.EntryNotFound">EntryNotFound</a></li>
<li><a title="sake.errors.InvalidDataFound" href="#sake.errors.InvalidDataFound">InvalidDataFound</a></li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="sake.errors.SakeException.base"><code class="name">var <span class="ident">base</span> :Â Optional[BaseException]</code></dt>
<dd>
<div class="desc"><p>The exception which caused this exception if applicable else <code>builtins.None</code>.</p></div>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="sake.errors.SakeException.base_exception"><code class="name">var <span class="ident">base_exception</span></code></dt>
<dd>
<div class="desc"><p>Return an attribute of instance, which is of type owner.</p></div>
</dd>
<dt id="sake.errors.SakeException.message"><code class="name">var <span class="ident">message</span> :Â str</code></dt>
<dd>
<div class="desc"><p>The exception's message, this may be an empty string if there is no message.</p></div>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="sake" href="index.html">sake</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="sake.errors.BackendError" href="#sake.errors.BackendError">BackendError</a></code></h4>
</li>
<li>
<h4><code><a title="sake.errors.CannotDelete" href="#sake.errors.CannotDelete">CannotDelete</a></code></h4>
</li>
<li>
<h4><code><a title="sake.errors.EntryNotFound" href="#sake.errors.EntryNotFound">EntryNotFound</a></code></h4>
</li>
<li>
<h4><code><a title="sake.errors.InvalidDataFound" href="#sake.errors.InvalidDataFound">InvalidDataFound</a></code></h4>
</li>
<li>
<h4><code><a title="sake.errors.SakeException" href="#sake.errors.SakeException">SakeException</a></code></h4>
<ul class="">
<li><code><a title="sake.errors.SakeException.base" href="#sake.errors.SakeException.base">base</a></code></li>
<li><code><a title="sake.errors.SakeException.base_exception" href="#sake.errors.SakeException.base_exception">base_exception</a></code></li>
<li><code><a title="sake.errors.SakeException.message" href="#sake.errors.SakeException.message">message</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.1</a>.</p>
</footer>
</body>
</html>